clouddrift/pairs.py

```diff
@@ -0,0 +1,407 @@
+"""
+Functions to analyze pairs of contiguous data segments.
+"""
+from clouddrift import ragged, sphere
+from concurrent.futures import as_completed, ThreadPoolExecutor
+import itertools
+import numpy as np
+import pandas as pd
+import xarray as xr
+from typing import Optional, Tuple, Union
+
+array_like = Union[list[float], np.ndarray[float], pd.Series, xr.DataArray]
+
+
+def chance_pair(
+    lon1: array_like,
+    lat1: array_like,
+    lon2: array_like,
+    lat2: array_like,
+    time1: Optional[array_like] = None,
+    time2: Optional[array_like] = None,
+    space_tolerance: Optional[float] = 0,
+    time_tolerance: Optional[float] = 0,
+):
+    """Given two arrays of longitudes and latitudes, return the chance of
+    finding a pair of points that overlap in space and time.
+
+    Parameters
+    ----------
+    lon1 : array_like
+        First array of longitudes in degrees.
+    lat1 : array_like
+        First array of latitudes in degrees.
+    lon2 : array_like
+        Second array of longitudes in degrees.
+    lat2 : array_like
+        Second array of latitudes in degrees.
+    time1 : array_like, optional
+        First array of times.
+    time2 : array_like, optional
+        Second array of times.
+    space_tolerance : float, optional
+        Tolerance in meters for the spatial overlap. If the overlap is within
+        this tolerance, the pair is considered to be a chance pair. Default is
+        0, or no tolerance.
+    time_tolerance : float, optional
+        Tolerance for the temporal overlap. If the overlap is within this
+        tolerance, and the space_tolerance condition is satisfied, the pair is
+        considered a chance pair.. Default is 0, or no tolerance.
+
+    Returns
+    -------
+    indices1 : np.ndarray[int]
+        Indices of the first array that satisfy chance pair criteria.
+    indices2 : np.ndarray[int]
+        Indices of the second array that satisfy chance pair criteria.
+
+    Examples
+    --------
+    In the following example, we load the first two trajectories from the GLAD
+    dataset and find all longitudes, latitudes, and times that satisfy the
+    chance pair criteria of 6 km separation distance and no time separation:
+
+    >>> from clouddrift.datasets import glad
+    >>> from clouddrift.pairs import chance_pair
+    >>> from clouddrift.ragged import unpack
+    >>> ds = glad()
+    >>> lon1 = unpack(ds["longitude"], ds["rowsize"], rows=0).pop()
```

Comment:
This I never use, it is clever. Could this be wrapped in a function eventually?

Reply:
I often forget about it. It could be useful as a convenience function to wrap this.